x-pack/plugins/ecs_data_quality_dashboard/server/routes/results/get_results.ts

```diff
@@ -0,0 +1,109 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { IRouter, Logger } from '@kbn/core/server';
+
+import { RESULTS_ROUTE_PATH, INTERNAL_API_VERSION } from '../../../common/constants';
+import { buildResponse } from '../../lib/build_response';
+import { buildRouteValidation } from '../../schemas/common';
+import { GetResultQuery } from '../../schemas/result';
+import type { Result, ResultDocument } from '../../schemas/result';
+import { API_DEFAULT_ERROR_MESSAGE } from '../../translations';
+import type { DataQualityDashboardRequestHandlerContext } from '../../types';
+import { createResultFromDocument } from './parser';
+import { API_RESULTS_INDEX_NOT_AVAILABLE } from './translations';
+
+export const getQuery = (patterns: string[]) => ({
+  size: 0,
+  query: {
+    bool: { filter: [{ terms: { 'rollup.pattern': patterns } }] },
+  },
+  aggs: {
+    latest: {
+      terms: { field: 'rollup.pattern', size: 10000 }, // big enough to get all patterns, but under `index.max_terms_count` (default 65536)
+      aggs: { latest_doc: { top_hits: { size: 1, sort: [{ '@timestamp': { order: 'desc' } }] } } },
+    },
+  },
+});
+
+export interface LatestAggResponseBucket {
+  key: string;
+  latest_doc: { hits: { hits: Array<{ _source: ResultDocument }> } };
+}
+
+export const getResultsRoute = (
+  router: IRouter<DataQualityDashboardRequestHandlerContext>,
+  logger: Logger
+) => {
+  router.versioned
+    .get({
+      path: RESULTS_ROUTE_PATH,
+      access: 'internal',
+      options: { tags: ['access:securitySolution'] },
+    })
+    .addVersion(
+      {
+        version: INTERNAL_API_VERSION,
+        validate: { request: { query: buildRouteValidation(GetResultQuery) } },
+      },
+      async (context, request, response) => {
+        const services = await context.resolve(['core', 'dataQualityDashboard']);
+        const resp = buildResponse(response);
+
+        let index: string;
+        try {
+          index = await services.dataQualityDashboard.getResultsIndexName();
+        } catch (err) {
+          logger.error(`[GET results] Error retrieving results index name: ${err.message}`);
+          return resp.error({
+            body: `${API_RESULTS_INDEX_NOT_AVAILABLE}: ${err.message}`,
+            statusCode: 503,
+          });
+        }
+
+        try {
+          // Confirm user has authorization for the requested patterns
+          const { patterns } = request.query;
+          const userEsClient = services.core.elasticsearch.client.asCurrentUser;
+          const privileges = await userEsClient.security.hasPrivileges({
+            index: [
+              { names: patterns.split(','), privileges: ['all', 'read', 'view_index_metadata'] },
+            ],
+          });
+          const authorizedPatterns = Object.keys(privileges.index).filter((pattern) =>
```

Comment:
Consider an alternative authorization check, because per the following scenario, this check appears to return unexpected results when two users visit the Data Quality page, but have different index privileges for the same _pattern_.

Example: given a pattern from the Security Solution default, e.g. `packetbeat-*`, one user may have privileges to view _all_ `packetbeat-*` data streams and indicies (i.e. `.ds-packetbeat-8.9.2-2023.12.06-000001` and `packetbeat-7.17.9-2024.01.08-000002`), while another user may have privileges for ONLY data streams created by the `8.x` version of Packetbeat (i.e. `.ds-packetbeat-8.9.2-2023.12.06-000001`), but not the `7.x` version of Packetbeat.

To demonstrate the unexpected results, first add the following `console.log` statements to the authorization step, such that the Kibana server logs include the `privileges` and derived `authorizedPatterns`:

```ts
          // Confirm user has authorization for the requested patterns
          const { patterns } = request.query;
          const userEsClient = services.core.elasticsearch.client.asCurrentUser;
          const privileges = await userEsClient.security.hasPrivileges({
            index: [
              { names: patterns.split(','), privileges: ['all', 'read', 'view_index_metadata'] },
            ],
          });

          console.log('--> privileges', JSON.stringify(privileges, null, 2));

          const authorizedPatterns = Object.keys(privileges.index).filter((pattern) =>
            Object.values(privileges.index[pattern]).some((v) => v === true)
          );

          console.log('--> authorizedPatterns', JSON.stringify(authorizedPatterns, null, 2));

          if (authorizedPatterns.length === 0) {
            return response.ok({ body: [] });
          }
```

Next, create a new role called `ONLY_SOME_PACKETBEAT_ROLE`, such that
- The user has access to _all_ `auditbeat-*` indices
- The user has access to only _some_ `packetbeat-*` indices

per the following screenshot of the `ONLY_SOME_PACKETBEAT_ROLE`:

![ONLY_SOME_PACKETBEAT_ROLE](https://github.com/elastic/kibana/assets/4459398/36b7c536-2318-4e1b-98d3-4ffb04db0e70)

In the screenshot above, the following pattern ONLY matches data streams created by the `8.x` version of Packetbeat:

```
packetbeat-8.*
```

As a result, users with the `ONLY_SOME_PACKETBEAT_ROLE` may view (and check), for example, a data stream created by the `8.x` version of Packetbeat, with a backing index named:

```
.ds-packetbeat-8.9.2-2023.12.06-000001
```

but that same user may NOT check an index created by the `7.x` version of Packetbeat, e.g. an index named:

```
packetbeat-7.17.9-2024.01.08-000002
```

Next create a user named `somepacketbeat`, and assign that user the  `ONLY_SOME_PACKETBEAT_ROLE`, per the following screenshot:

![somepacketbeat_user](https://github.com/elastic/kibana/assets/4459398/20510359-2e76-49e0-bd76-ea0f543c861a)

Finally, create one more user named `alsosuperuser`, and assign that user the `superuser` role, per the screenshot below:

![alsosuperuser](https://github.com/elastic/kibana/assets/4459398/e4e88840-84da-41e7-8711-d3644daacf2a)

Navigate to Stack Management > Users. Per the screenshot below:

![users](https://github.com/elastic/kibana/assets/4459398/c682076c-cc79-4f59-ad7e-9e97a836c225)

✅ The `alsosuperuser` user has the `superuser` role (just like the `elastic` user)
✅ The `somepacketbeat` user has the `ONLY_SOME_PACKETBEAT_ROLE`

### Establish a baseline with the `elastic` user

1) Login to Kibana as the `elastic` user, who has the `superuser` role

2) Navigate to Security > Dashboards > Data Qualty

3) Click the `Check all` button

Note that:

✅  The `.alerts-security.alerts-default` pattern contains results, per the screenshot below

![elastic_user_alerts_and_auditbeat_results](https://github.com/elastic/kibana/assets/4459398/7665a625-92dd-4e0a-9b77-501d426b9b59)

✅ The `auditbeat-*` pattern contains results, per previous screenshot above

✅  The `packetbeat-*` pattern contains `6` pages of results, which include indices from both the `8.x` and `7.x` versions of Packetbeat, per the screeenshot below:

![elastic_packetbeat_8_x_and_7x](https://github.com/elastic/kibana/assets/4459398/d34eead2-5ccb-47a5-bf38-cd2df1ce4922)

Also observe the Kibana server logs `console.log` output for the `elastic` user:

```
--> privileges {
  "username": "elastic",
  "has_all_requested": true,
  "cluster": {},
  "index": {
    ".alerts-security.alerts-default": {
      "all": true,
      "read": true,
      "view_index_metadata": true
    },
    "auditbeat-*": {
      "all": true,
      "read": true,
      "view_index_metadata": true
    },
    "logs-*": {
      "all": true,
      "read": true,
      "view_index_metadata": true
    },
    "packetbeat-*": {
      "all": true,
      "read": true,
      "view_index_metadata": true
    }
  },
  "application": {}
}
--> authorizedPatterns [
  ".alerts-security.alerts-default",
  "auditbeat-*",
  "logs-*",
  "packetbeat-*"
]
```

### Compare the `alsosuperuser` user with the `elastic` user

1) Login to Kibana as the `alsosuperuser` user, who also has the `superuser` role, just like the `elastic` user

2) Navigate to Security > Dashboards > Data Quality

Note that:

✅  As expected, `Last checked: --` is displayed for the `alsosuperuser` user, because this user has never performed a data quality check

✅  As expected, the `.alerts-security.alerts-default` pattern contains results, because the `alsosuperuser` is viewing results created by the `elastic` user,  per the screenshot below:

![alsosuperuser_alerts_and_auditbeat_results](https://github.com/elastic/kibana/assets/4459398/c8e467af-68c9-47c3-b90f-cbf6d81b38e0)

✅ As expected, the `auditbeat-*` pattern contains results (from the `elastic` user), per previous screenshot above

✅  As expected, the `packetbeat-*` pattern contains `6` pages of results (created by the `elastic` user), which include indices from both the `8.x` and `7.x` versions of Packetbeat, per the screeenshot below:

![alsosuperuser_packetbeat_8_x_and_7x](https://github.com/elastic/kibana/assets/4459398/ef5f6b7a-38bb-4454-8f19-72ffe892e06b)

Also observe the Kibana server logs `console.log` output for the `alsosuperuser` user:

```
--> privileges {
  "username": "alsosuperuser",
  "has_all_requested": true,
  "cluster": {},
  "index": {
    ".alerts-security.alerts-default": {
      "all": true,
      "read": true,
      "view_index_metadata": true
    },
    "auditbeat-*": {
      "all": true,
      "read": true,
      "view_index_metadata": true
    },
    "logs-*": {
      "all": true,
      "read": true,
      "view_index_metadata": true
    },
    "packetbeat-*": {
      "all": true,
      "read": true,
      "view_index_metadata": true
    }
  },
  "application": {}
}
--> authorizedPatterns [
  ".alerts-security.alerts-default",
  "auditbeat-*",
  "logs-*",
  "packetbeat-*"
]
```

So far, the output demonstrates that:

✅ As expected, the `alsosuperuser` user may view all the persisted results from the `elastic` user

✅ As expected, the `privileges` and derived `authorizedPatterns` are the same for both the `alsosuperuser` and `elastic` user

### Compare the `somepacketbeat` user with the `elastic` user (has unexpected results)

1) Login to Kibana as the `somepacketbeat` user, who has DIFFERENT permissions compared with the `elastic` and `alsosuperuser` users

2) Navigate to Security > Dashboards > Data Quality

Note that:

✅  As expected, `Last checked: --` is displayed for the `somepacketbeat` user, because this user has never performed a data quality check, per the screeenshot below

![somepacketbeat_data_quality](https://github.com/elastic/kibana/assets/4459398/960ead16-68f5-4be3-8ff7-14e82707f95f)

✅  As expected, the `.alerts-security.alerts-default` pattern is not displayed, because the `somepacketbeat` user (unlike the `elastic` user) does not have access to the `.alerts-security.alerts-default` pattern

✅ As expected, the `auditbeat-*` pattern contains results (from the `elastic` user)

✅  As expected, the `packetbeat-*` pattern contains `3` pages of indices, (as opposed to `6` for the `elastic` and `alsosuperuser` users), because the `ONLY_SOME_PACKETBEAT_ROLE` role only allows access to data streams created by the `8.x` version of Packetbeat

❌ NOT EXPECTED: The `packetbeat-*` results (created by the `elastic` user) are NOT shown for the `somepacketbeat` user, even though the `somepacketbeat` has access to the `8.x` Packetbeat data streams

This unexpected behavior appears to be related to the following unexpected `authorizedPatterns` derived state in the Kibana server log for the `somepacketbeat` user:

```
--> privileges {
  "username": "somepacketbeat",
  "has_all_requested": false,
  "cluster": {},
  "index": {
    "auditbeat-*": {
      "all": false,
      "read": true,
      "view_index_metadata": true
    },
    "packetbeat-*": {
      "all": false,
      "read": false,
      "view_index_metadata": false
    }
  },
  "application": {}
}
--> authorizedPatterns [
  "auditbeat-*"
]
```

❌ NOT EXPECTED: The `authorizedPatterns` in the Kibana server log output for the `somepacketbeat` user does NOT include Packeatbeat, even though the `stats` API (correctly) returns results for data streams created by the `8.x` version of packetbeat.

It appears that Packetbeat is not included in `authorizedPatterns` because the `privileges` entry for Packetbeat in the `console.log` output:

```
    "packetbeat-*": {
      "all": false,
      "read": false,
      "view_index_metadata": false
    }
  },
```

has `"read": false` and `"view_index_metadata": false`.

It appears both `read` and `view_index_metadata` are `false` for the `somepacketbeat` user because the `packetbeat-*` pattern matches data streams created by the 8.x version of Packetbeat (e.g. `.ds-packetbeat-8.9.2-2023.12.06-000001`), which the user has access to, AND indices created by the 7.x version (e.g. `packetbeat-7.17.9-2024.01.08-000002`), which the `somepacketbeat` user does NOT have access to.

Conclusion: it appears that the current implementation of authorization returns incorrect results when the specified pattern is not applicable to ALL data streams and / or indices that match the pattern.